pyop2/codegen/rep2loopy.py

```diff
@@ -85,6 +91,210 @@ def petsc_function_lookup(target, identifier):
+
+        super(SolveCallable, self).__init__(name,
+                                            arg_id_to_dtype=arg_id_to_dtype,
+                                            arg_id_to_descr=arg_id_to_descr)
+
+        self.name = name
+        self.name_in_target = name_in_target
+
+    def with_types(self, arg_id_to_dtype, kernel, callables_table):
+        for i in range(len(arg_id_to_dtype)):
+            if arg_id_to_dtype.get(i) is None:
+                # the types provided aren't mature enough to specialize the
+                # callable
+                return (self.copy(arg_id_to_dtype=arg_id_to_dtype),
+                        callables_table)
+
+        mat_dtype = arg_id_to_dtype[0].numpy_dtype
+        name_in_target = "solve"
+
+        return (self.copy(name_in_target=name_in_target,
+                arg_id_to_dtype={-1: NumpyType(mat_dtype, target=loopy.CTarget()), 0: NumpyType(mat_dtype, target=loopy.CTarget()), 1: NumpyType(mat_dtype, target=loopy.CTarget()), 2: NumpyType(numpy.int32, target=loopy.CTarget())}),
+                callables_table)
+
+    def emit_call_insn(self, insn, target, expression_to_code_mapper):
+        assert self.is_ready_for_codegen()
+        assert isinstance(insn, loopy.CallInstruction)
+
+        parameters = insn.expression.parameters
+
+        if type(parameters) != list:
+            parameters = list(parameters)
+        par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+        parameters.append(insn.assignees[0])
+        par_dtypes.append(self.arg_id_to_dtype[1])
+        mat_descr_A = self.arg_id_to_descr[0]
+
+        arg_c_parameters = [
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            if isinstance(par, SubArrayRef) else
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            for par, par_dtype in zip(parameters, par_dtypes)
+        ]
+        c_parameters = []
+        c_parameters.insert(0, arg_c_parameters[-1])  # out
+        c_parameters.insert(1, arg_c_parameters[0])  # A
+        c_parameters.insert(2, arg_c_parameters[1])  # B
+        c_parameters.insert(3, mat_descr_A.shape[1])  # n
+        return var(self.name_in_target)(*c_parameters), False
+
+    def generate_preambles(self, target):
+        assert isinstance(target, CTarget)
+        code = """#include <string.h>
+            #include <stdio.h>
+            #include <stdlib.h>
+            #ifndef Solve_HPP
+            #define Solve_HPP
+            #define BUF_SIZE 30
+
+            static PetscBLASInt ipiv_buffer[BUF_SIZE];
+            static void solve(PetscScalar* restrict out, const PetscScalar* restrict A, PetscScalar* restrict B, PetscBLASInt N)
```

Comment:
Prefer `__restrict__` over `restrict`.

In both cases, I think `N` is compile-time known, so can we just inline it, rather than passing as a parameter?

Reply:
N is inherited earlier from the size of the matrix A (line 235 + 257), so you do not need to define N in the according loopy call instruction. Why would it be better to inline?
And with inlining you mean something like `const PetscBLASInt N = sizeof((A))/sizeof((A)[0]);` inside the C function defined here, or do you mean generating a loopy statement for N in the slate kernel.

Reply:
So since this preamble is generated for a problem with a compile-time known N, we can have two versions:

1. no runtime memory allocation occurs and the code looks something like
```c

static void solve(PetscScalar* __restrict__ out, const PetscScalar * __restrict__ A, const PetscScalar * __restrict__ B)
{
   PetscBLASInt ipiv[literal_value_of_N];
   memcpy(out,B,N*sizeof(*out));
  const PetscBLASInt NRHS = 1;
  const PetscBLASInt N = literal_value_of_N;
  LAPACKgesv_(&N, &NRHS, A, &N, ipiv, out, &N, &info);
  ...
}
```

And another version where we have:

```
static void solve(PetscScalar* __restrict__ out, const PetscScalar * __restrict__ A, const PetscScalar * __restrict__ B)
{
   PetscBLASInt *ipiv = malloc(literal_value_of_N * sizeof(*ipiv));
   memcpy(out,B,N*sizeof(*out));
  const PetscBLASInt NRHS = 1;
  const PetscBLASInt N = literal_value_of_N;
  LAPACKgesv_(&N, &NRHS, A, &N, ipiv, out, &N, &info);
  ...
  free(ipiv);
}
```

Reply:
Where is literal_value_of_N coming from?

Reply:
Ah, I had thought we make these things when `N` is readily available to us, but perhaps it is not. If so, ignore me.